File: test/ModularPipelines.UnitTests/Helpers/ToolResolverTests.cs
using ModularPipelines.Attributes;
using ModularPipelines.Helpers.Internal;
using ModularPipelines.Options;

namespace ModularPipelines.UnitTests.Helpers;

public class ToolResolverTests
{
    public async Task ResolveTool_FromDirectAttribute_ReturnsTool()
    {
        var resolver = new ToolResolver();

        var tool = resolver.ResolveTool(typeof(DirectToolOptions));

        await Assert.That(tool).IsEqualTo("mytool");
    }

    public async Task ResolveTool_FromInheritedAttribute_ReturnsTool()
    {
        var resolver = new ToolResolver();

        var tool = resolver.ResolveTool(typeof(InheritedToolOptions));

        await Assert.That(tool).IsEqualTo("git");
    }

    public async Task ResolveTool_FromDeeplyInheritedAttribute_ReturnsTool()
    {
        var resolver = new ToolResolver();

        var tool = resolver.ResolveTool(typeof(DeeplyInheritedOptions));

        await Assert.That(tool).IsEqualTo("docker");
    }

    public async Task ResolveTool_NoAttribute_ReturnsNull()
    {
        var resolver = new ToolResolver();

        var tool = resolver.ResolveTool(typeof(NoToolOptions));

        await Assert.That(tool).IsNull();
    }

    public async Task ResolveTool_FromInstance_ReturnsToolFromAttribute()
    {
        var resolver = new ToolResolver();
        var options = new DirectToolOptions();

        var tool = resolver.ResolveTool(options);

        await Assert.That(tool).IsEqualTo("mytool");
    }

    // Test fixtures
    [CliTool("mytool")]
    private record DirectToolOptions : CommandLineToolOptions;

    [CliTool("git")]
    private abstract record GitOptionsBase : CommandLineToolOptions;

    private record InheritedToolOptions : GitOptionsBase;

    [CliTool("docker")]
    private abstract record DockerOptionsBase : CommandLineToolOptions;

    private abstract record DockerContainerOptions : DockerOptionsBase;

    private record DeeplyInheritedOptions : DockerContainerOptions;

    private record NoToolOptions : CommandLineToolOptions;
}

File: test/ModularPipelines.UnitTests.FSharp/Helpers/ToolResolverTests.fs
namespace ModularPipelines.UnitTests.FSharp.Helpers

open ModularPipelines.UnitTests.Helpers
open ModularPipelines.UnitTests.FSharp
open TUnit.Core

type ToolResolverTests() =
    inherit ModularPipelines.UnitTests.Helpers.ToolResolverTests()

    [<Test>]
    member this.ResolveTool_FromDirectAttribute_ReturnsTool() =
        CSharpTestWrapper.invokeTest this typeof<ModularPipelines.UnitTests.Helpers.ToolResolverTests> "ResolveTool_FromDirectAttribute_ReturnsTool" None

    [<Test>]
    member this.ResolveTool_FromInheritedAttribute_ReturnsTool() =
        CSharpTestWrapper.invokeTest this typeof<ModularPipelines.UnitTests.Helpers.ToolResolverTests> "ResolveTool_FromInheritedAttribute_ReturnsTool" None

    [<Test>]
    member this.ResolveTool_FromDeeplyInheritedAttribute_ReturnsTool() =
        CSharpTestWrapper.invokeTest this typeof<ModularPipelines.UnitTests.Helpers.ToolResolverTests> "ResolveTool_FromDeeplyInheritedAttribute_ReturnsTool" None

    [<Test>]
    member this.ResolveTool_NoAttribute_ReturnsNull() =
        CSharpTestWrapper.invokeTest this typeof<ModularPipelines.UnitTests.Helpers.ToolResolverTests> "ResolveTool_NoAttribute_ReturnsNull" None

    [<Test>]
    member this.ResolveTool_FromInstance_ReturnsToolFromAttribute() =
        CSharpTestWrapper.invokeTest this typeof<ModularPipelines.UnitTests.Helpers.ToolResolverTests> "ResolveTool_FromInstance_ReturnsToolFromAttribute" None
